### DISCUSSION TITLE
Cannot reload fonts on DX12

### DISCUSSION THREAD

**Discussion by nr-jpg** [ocornut/imgui#8174] [New Users Build/Link/Run issues ONLY!]
Howdy, I'm trying to do some basic DPI awareness on a DX12 app by reloading the fonts as starters. My code logic in a nutshell is to wait for the GPU to be idle if there is a pending DPI awareness change (from WM_DPICHANGED message) and then clear the fonts and reload them with the new scale

    WaitForGPU();
    
    const float dpiScale = std::max(GameWindow::GetInstance().GetDPI() / 96.0f, 1.0f);
    const float fontScale = dpiScale * GlobalFontScale;
    
    ImGui::GetIO().Fonts->ClearFonts();
    ImGui::GetIO().Fonts->AddFontFromFileTTF(
        "C:\\Windows\\Fonts\\Arial.ttf",
        14.0f * fontScale,
        NULL,
        ImGui::GetIO().Fonts->GetGlyphRangesDefault());

I call the above code after the present, I use the same logic for resizing, fullscreen, etc, no issues there.

I call the same code during initialization for the initial font load and for the starting window imgui works just fine. However, attempting to move to a monitor with different DPI causes the following assert:
  
    Assertion failed!
    
    Program: ...\Binaries\Debug\playground.exe
    File: ...\imgui-1...\imgui.cpp
    Line: 10344
    
    Expression: g.IO.Fonts->IsBuilt() && "Font Atlas not built! Make sure you called ImGui_ImplXXXX_NewFrame() function for renderer backend, which should call io.Fonts->GetTexDataAsRGBA32() / GetTexDataAsAlpha8()"

Looking at the code the assert message is misleading/wrong, new frame for DX12 does nothing if the device objects exist, so I tried adding a "ImGui_ImplDX12_InvalidateDeviceObjects()" call as well when reloading fonts. This results in another assert during the "ImGui_ImplDX12_NewFrame()" or "create device objects" call:

    Assertion failed!
    
    Program: ...\Debug\playground.exe
    File: ...\im...\imgui_draw.cpp
    Line: 2827
    
    Expression: cfg.DstFont && (!cfg.DstFont->IsLoaded() || cfg.DstFont->ContainerAtlas == atlas)

Any ideas what I'm missing here?

I was looking at this post + FAQ, but did not see anything about the issues I'm having.
https://github.com/ocornut/imgui/discussions/3925

Using imgui-1.91.5.

**Comment by ocornut** [comment]
(I’ll try to look but FYI i am in parallel rewriting the system to facilitate more dynamic use of fonts and incremental texture updates, so hopefully there will be a much better version of this in a month or two)

**Reply by ocornut** [reply]
But FYI it should already be possible to do what you are doing, perhaps providing code would help narrowing your issue.

**Reply by nr-jpg** [reply]
Thanks for your reply, I'm still using the demo view to get it working properly, so there is not much code to provide really.

The init:

            IMGUI_CHECKVERSION();

            ImGui::CreateContext();
            Bool status = ImGui_ImplWin32_Init(GameWindow::GetInstance().GetHandle());
            Assert(status);

            D3D12_DESCRIPTOR_HEAP_TYPE heapType = D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV;
            D3D12_CPU_DESCRIPTOR_HANDLE imGuiFontCPUHandle = GetGPUDescriptorAllocator()->AllocPersistent(heapType, 1);
            D3D12_GPU_DESCRIPTOR_HANDLE imGuiFontGPUHandle = GetGPUDescriptorAllocator()->GetGPUHandle(imGuiFontCPUHandle, heapType);

            ImGui_ImplDX12_Init(
                GetDevice(),
                Config::Graphic::MaxFramesInProgress,
                Config::Graphic::SwapChainFormat,
                GetGPUDescriptorAllocator()->GetHeap(heapType),
                imGuiFontCPUHandle,
                imGuiFontGPUHandle);

            UpdateImGuiFonts(); // contains the code from above with font reload
            
The update + rendering part (I've copy-pasted them together for now)

        ImGui_ImplDX12_NewFrame();
        ImGui_ImplWin32_NewFrame();
        ImGui::NewFrame();
        ImGui::ShowDemoWindow(); // Show demo window! :)

        RenderTarget* backBuffer = GetSwapChain()->GetCurrentFrameBuffer();
        context.Transition(backBuffer, D3D12_RESOURCE_STATE_RENDER_TARGET);

        auto cpuHandle = backBuffer->GetRTV().GetCPUHandle();
        context->OMSetRenderTargets(1, &cpuHandle, false, nullptr);

        ImGui::Render();
        ImGui_ImplDX12_RenderDrawData(ImGui::GetDrawData(), context.GetD3DCommandList());

**Reply by nr-jpg** [reply]
Seems like it's easily reproductible in the example_win32_directx12 project.

Inserting the following code before the NewFrame calls produces same error:

        WaitForLastSubmittedFrame();

        ImGui_ImplDX12_InvalidateDeviceObjects();

        ImGui::GetIO().Fonts->ClearFonts();
        ImGui::GetIO().Fonts->AddFontFromFileTTF(
            "C:\\Windows\\Fonts\\Arial.ttf",
            14.0f,
            NULL,
            ImGui::GetIO().Fonts->GetGlyphRangesDefault());

**Reply by sergeyn** [reply]
you can check my branch, I think I got it correct for both dx11 and dx12

**Comment by ocornut** [comment]
Sorry for my late answer.
Changing
```cpp
ImGui::GetIO().Fonts->ClearFonts();
```
to
```cpp
ImGui::GetIO().Fonts->Clear();
```

Should fix it.

Going to look why ClearFonts() create this situation, but my expectation that is the individual ClearXXX() function will be removed anyhow.

**Reply by nr-jpg** [reply]
That did it, thanks! It could also be nice to include a DPI awareness sample in the demo.

**Comment by ocornut** [comment]
Basically those functions `ClearInputData()` `ClearTexData()` `ClearFonts()` are not well formalized, and calling ClearFonts() without `ClearInputData()` is causing an issue. I will end up removing one soon.

**Reply by ocornut** [reply]
Pushed a fix 1019934

Basically `ClearFonts()` has a misleading name, since it clears output fonts, not the data added by `AddFont()`.
It's simpler to clear them both.
